src/Microsoft.Extensions.Caching.Memory/MemoryCacheServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Extension methods for setting up memory cache related services in an <see cref="IServiceCollection" />.
+    /// </summary>
+    public static class MemoryCacheServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds memory caching services to the specified <see cref="IServiceCollection" />.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
+        public static void AddMemoryCache(this IServiceCollection services)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            services.AddOptions();
+            services.TryAdd(ServiceDescriptor.Transient<IDistributedCache, MemoryDistributedCache>());
```

Comment:
Memory cache and MemoryDistributedCache are always required for our system to work as far as I understood. That's the reason why they are bundled together and not done as two separate calls? (The other option we contemplated). I really want to avoid making the user make two calls to extension methods on IServiceCollection to configure the system correctly (it's very easy to forget one).

Another option, although @lodejard doesn't like it very much, it's to have an ICachingBuilder and make AddRedisCache and AddSqlServerCache handle from there. That way a user would do something like

```
services
    .AddCaching()
    .UseRedisCache(ro => {})
```

As far as I understood, what @lodejard doesn't like about this pattern is that we are technically not extending the system but changing how it works. Maybe you feel different about it.

In any case, the other changes (Make AddRedisCache and AddSqlServerCache call Add instead of TryAdd) are still useful as they take away the order dependency on how you call the methods.
(AddCaching().AddRedisCache() would not work)

Also, caching is one of those things I would resolve with a factory. If you need different caches for different parts of the system, a factory is more flexible that just using pure DI. You can inject a factory wherever your cache is needed and the factory can have a create method with parameters that you can use to pick the cache type most appropriate to your component.

So after all this, I think we have 4 options:
* Keep everything as it is right now.
* Use what I have in the current PR.
* Use the builder pattern I proposed as alternative.
* Split AddCaching into AddLocalCache() and AddMemoryCache() (we can add the DistributedSuffix to Add(Memory|Redis|SqlServer)Cache.

In my opinion 2 and 3 are ok.

@Eilon what is your take?